### PR TITLE
Eliminate BigInt allocs for 128-bit FixedDecimal conversions.

### PR DESCRIPTION
**Pull request by NHDaly**
Before this PR, converting an Int to a FD{Int128} had to allocate a BigInt, which is fairly expensive:
```julia
julia> @time convert(FixedDecimal{Int128, 2}, 1)
  0.000003 seconds (7 allocations: 128 bytes)
FixedDecimal{Int128,2}(1.00)
```
After, the allocs are gone, and this is only involves the checks for overflow and the coefficient multiplication:
```julia
julia> @time convert(FixedDecimal{Int128, 2}, 1)
  0.000000 seconds
FixedDecimal{Int128,2}(1.00)
```

I've added tests to cover all corner cases and to ensure the allocations remain eliminated.

Also, while I was here, I have changed the InexactErrors you get for illegal conversions to be clearer, by reporting on the types involved in the conversion directly, rather than the internal errors you got previously:
```julia
Before: InexactError(:Int128, Int128, 34028236692093846346337460743176821145500)
 After: InexactError(:convert, FixedDecimal{Int128, 2}, 0xffffffffffffffffffffffffffffffff)

Before: InexactError: Int128(17014118346046923173168730371588410572700)
 After: InexactError: convert(FixedDecimal{Int128, 2}, 170141183460469231731687303715884105727)
```

Thanks!

### CODE REVIEW

File: src/FixedPointDecimals.jl [hunk 1]
```diff
@@ -296,7 +296,35 @@ end
...
+    typemin(T) <= x <= typemax(T) || throw_inexact()
+    xT = convert(T, x)  # This won't throw, since we already checked, above.
+    # Perform x * C, and check for overflow. This is cheaper than a widemul, especially for
+    # 128-bit T, since those widen into a BigInt.
+    v = _mul_checked_overflow(throw_inexact, xT, C)
+    reinterpret(FD{T, f}, v)
+end
+function Base.convert(::Type{FD{BigInt, f}}, x::Integer) where {f}
+    # We specialize on f==1, since julia can't eliminate BigInt multiplication.
+    if f == 1
```
Comment (Drvi) [inline src/FixedPointDecimals.jl:310]: Sorry, previously I said we can specialize the case when `f == 1` but I meant when the `coefficient` is one, so when `f == 0`

**Review by Drvi — COMMENTED** [review]
I'll have a look at the tests on Monday

File: src/FixedPointDecimals.jl (hunk 1 above)
Comment (NHDaly) [reply src/FixedPointDecimals.jl:310]: Oh, heh, whoops! I meant the same thing (as is evident by using the BigInt value unchanged). Duh, of course it should have been f=0. Sorry I didn't catch that myself! Also, I guess I didn't add test coverage for this case 😓 I added in the BigInt handling at the end and I wasn't being as thorough then, I think. :(

Comment (NHDaly) [reply src/FixedPointDecimals.jl]: Thanks Tomas. Fixed this in the last commit.

Comment (Drvi) [reply src/FixedPointDecimals.jl]: Thanks! Also, the comment needs updating.

File: test/FixedDecimal.jl [hunk 2]
```diff
@@ -0,0 +1,1098 @@
...
+    @test @allocated(convert(F64D2, Int8(-1))) == 0
+    @test @allocated(convert(F64D2, UInt8(1))) == 0
+    @test @allocated(convert(SFD2, Int8(-1))) == 0
+
+    # Int 128 cases
+    @test @allocated(convert(WFD2, 1)) == 0
+    @test @allocated(convert(WFD2, 1)) == 0
+    @test @allocated(convert(UWFD2, 1)) == 0
+    @test @allocated(convert(WFD2, -1)) == 0
+    # @test @allocated(convert(UWFD2, -1)) == 32   # not sure how best to test this
```
Comment (Drvi) [inline test/FixedDecimal.jl:363]: This should help https://github.com/JuliaLang/AllocCheck.jl

**Review by Drvi — APPROVED** [review]
LGTM!

File: test/FixedDecimal.jl (hunk 2 above)
Comment (NHDaly) [reply test/FixedDecimal.jl:363]: How so? Can you test there that is a single expected allocation? I didn't see that yet.

I did try this but it didn't work:
```julia
julia> using FixedPointDecimals

julia> const WFD2 = FixedDecimal{Int128, 2}
FixedDecimal{Int128, 2}

julia> @check_allocs f(x) = convert(WFD2, x)
f (generic function with 2 methods)

julia> f(-1)
ERROR: ReadOnlyMemoryError()
```

Comment (nickrobinson251) [reply test/FixedDecimal.jl:363]: > Can you test there that is a single expected allocation?


```suggestion
    @test @allocations(convert(UWFD2, -1)) == 1
```

Comment (Drvi) [reply test/FixedDecimal.jl:363]: It seems to work on 1.9:
```julia
julia> using AllocCheck

julia> @check_allocs myconv(::Type{T}, x) where T = convert(T, x)
myconv (generic function with 1 method)

julia> using FixedPointDecimals

julia> const UWFD2 = FixedDecimal{UInt128, 2}
FixedDecimal{UInt128, 2}

julia> myconv(UWFD2, -1) # no error
FixedDecimal{UInt128,2}(1125651476239148520299089100.85)

julia> b = big(1)
1

julia> myconv(UWFD2, b)
ERROR: @check_alloc function encountered 6 errors (6 allocations / 0 dynamic dispatches).
```
Not sure about the `ReadOnlyMemoryError` I get on 1.10

Comment (NHDaly) [reply test/FixedDecimal.jl:363]: aha that's right, i forgot this one throws; i wasn't looking very closely this morning. Thanks. I'm just not going to test this since it's the exception case. I'll just delete this line

Comment (NHDaly) [reply test/FixedDecimal.jl:363]: But yeah you're right, if this was working, it would be a good way to test it!

Comment (Drvi) [reply test/FixedDecimal.jl]: 👍
